### DISCUSSION TITLE
Feedback: Experimental Open API 3.1 support

### DISCUSSION THREAD

**Discussion by Stranger6667** [schemathesis/schemathesis#1822] [Q&A]
We've introduced experimental support for OpenAPI 3.1, focusing initially on compatible JSON Schema validation for responses. Data generation still adheres to OpenAPI 3.0.

### How to Enable

- CLI: `--experimental=openapi-3.1`
- Python API: `schemathesis.experimental.openapi_3_1.enable()`

We'd love to hear your feedback:

- Did it work as you expected?
- Did you encounter any issues?
- Any other thoughts or suggestions?

Your input is vital for refining this feature. Thank you for contributing to its development!

**Comment by hutchins** [comment]
CLI --experimental=openapi-3.1 works great, however I'm not clear how to implement the schemathesis.experimental.openapi_3_1.enable() in pytest code. Can you give a little more detail or an example?

**Reply by Stranger6667** [reply]
Sure! I am going to update the docs soon :)

In order to enable this in python tests you need to add this to your conftest.py (note that it enables open api 3.1 support globally):

```python
import schemathesis 
import pytest 

@pytest.fixture(autouse=True, scope=“session”)
def openapi_3_1_support():
    schemathesis.experimental.openapi_3_1.enable()

```

I think isolating this in a fixture is the way to go

**Reply by Stranger6667** [reply]
Hi @hutchins , I think I was too fast with providing the code sample that utilises pytest fixtures. There are cases when users place their schema loading logic to the module level, then it will happen earlier than the fixture is evaluated. Therefore this advice only works for lazy schema loading via `schemathesis.from_pytest_fixture`. So, the most reliable way to enable support in python tests would be calling `schemathesis.experimental.openapi_3_1.enable()` early in the test setup code, for example, the beginning of the root `conftest.py` file

**Reply by hutchins** [reply]
I'm using v3.20.1 and have tried putting `schemathesis.experimental.openapi_3_1.enable() ` in conftest or immediately after import and get AttributeError: module 'schemathesis.experimental' has no attribute 'openapi_3_1'.

Is there another way to use this new feature?

**Reply by hutchins** [reply]
% `export SCHEMATHESIS_EXPERIMENTAL_OPENAPI_3_1=true`
Try testing...
schemathesis.exceptions.SchemaError: The provided schema uses Open API 3.1.0, which is currently not supported.

When I try to get this to work in the module it doesn't work either. Steps to reproduce this error:

% cat test_me.py
import schemathesis
print(schemathesis.__version__)
schemathesis.experimental.openapi_3_1.enable()

% python3.11 test_me.py
3.20.1
Traceback (most recent call last):
  File ".../test_me.py", line 5, in <module>
    schemathesis.experimental.openapi_3_1.enable()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'schemathesis.experimental' has no attribute 'openapi_3_1'

**Reply by Stranger6667** [reply]
I am very sorry @hutchins - it is oversight from me, will fix it soon

**Reply by Stranger6667** [reply]
Fixed in `3.20.2`

**Reply by graipher** [reply]
I am still getting the `AttributeError`, even with `3.20.2`.

Using the env variable works just fine, though.

**Reply by Stranger6667** [reply]
@graipher Can you, please, share your code, pls? Maybe I didn't cover some usage pattern

**Reply by graipher** [reply]
I get the error with the same three lines shown above:

```
Python 3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import schemathesis
>>> print(schemathesis.__version__)
3.20.2
>>> schemathesis.experimental.openapi_3_1.enable()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'schemathesis.experimental' has no attribute 'openapi_3_1'
```

(and I get the same error when trying to add this to my test files as first line after the imports).

**Reply by Stranger6667** [reply]
Thanks! The original doc was incorrect, the attribute is named `OPEN_API_3_1`, I updated the [docs](https://schemathesis.readthedocs.io/en/stable/experimental.html#in-python-tests) but forgot to mention it here:

```python
import schemathesis

# Globally enable OpenAPI 3.1 experimental feature
schemathesis.experimental.OPEN_API_3_1.enable()
```

**Reply by hutchins** [reply]
I have found what appears to be an issue or it could be that I'm not using the python module correctly.
When using the module in pytest and OpenAPI 3.1 it reports failures that the CLI doesn't.

The consistent error found appears to be related to:

- Received a response with 'text/plain' Content-Type, but it is not declared in the schema.
- Defined content types: application/json
- Response payload: `Bad Request`

This is a different behavior then when using the CLI.

Code used to reproduce:

```
schemathesis.experimental.OPEN_API_3_1.enable()
schema = schemathesis.from_uri(
    url
    skip_deprecated_operations=True,
    headers={
        "Authorization": "Bearer " + token,
        "data-partition-id": data_partition,
    },
    code_sample_style="curl",
)
@schema.parametrize()
def test_openapi(case):
    headers = {
        "Authorization": "Bearer " + token,
        "data-partition-id": data_partition,
    }
    case.call_and_validate(headers=headers)
```

`st run -H "Authorization: Bearer xyz" -H "data-partition-id: xyz" --skip-deprecated-operations --code-sample-style curl --experimental=openapi-3.1 URL`

**Reply by Stranger6667** [reply]
I think that in CLI you need to specify `--checks=all` in order to run all the checks. But I missed that it is inconsistent between CLI & Python tests, thank you for reporting it!

**Comment by alan-cd** [comment]
Hi @Stranger6667 , we are using the Schemathesis for our API testing. But I met a problem with openapi-3.1 and concurrent testing.
Here is my command:
```
st run --workers auto --checks all *** --base-url *** --experimental=openapi-3.1 --hypothesis-suppress-health-check=filter_too_much
```
Initially I didn't add the parameter `--workers auto`, it works well. But when I add it, it throws a runtime error like below:
Here is the stacktrace:
```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/runner/impl/core.py", line 737, in network_test
    response = _network_test(case, *args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/runner/impl/core.py", line 784, in _network_test
    run_checks(
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/runner/impl/core.py", line 610, in run_checks
    skip_check = check(response, copied_case)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/specs/openapi/checks.py", line 129, in response_schema_conformance
    return case.operation.validate_response(response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/models.py", line 862, in validate_response
    return self.schema.validate_response(self, response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/specs/openapi/schemas.py", line 613, in validate_response
    if self.spec_version.startswith("3.1") and experimental.OPEN_API_3_1.is_enabled:
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/experimental/__init__.py", line 25, in is_enabled
    return self._storage.is_enabled(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/experimental/__init__.py", line 82, in is_enabled
    return feature in self.enabled
                      ^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/schemathesis/experimental/__init__.py", line 50, in enabled
    return self._local_data.enabled
           ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: '_thread._local' object has no attribute 'enabled'
```
I checked the trace log and I think it may relate to the experimental feature `OpenAPI-3.1`, so I comment here.

Do I make any mistakes?

**Reply by Stranger6667** [reply]
Hi, it looks like a bug, I’ll check it. Thanks for reporting

**Reply by Stranger6667** [reply]
Fixed, I am going to make a new release in the next few days

**Reply by Stranger6667** [reply]
Released in `3.25.2`

**Reply by alan-cd** [reply]
Great! I've just run this again and now it is passed. Thank you very much for creating this great tool!

**Reply by Stranger6667** [reply]
Thank you @alan-cd ! Happy to hear it from you :) do not hesitate to reach out if you have any other questions :)
